android-components/components/browser/state/src/main/java/mozilla/components/browser/state/engine/middleware/SessionPrioritizationMiddleware.kt

```diff
@@ -60,21 +100,38 @@ class SessionPrioritizationMiddleware : Middleware<BrowserState, BrowserAction>
         }
     }
 
-    private fun updatePriorityIfNeeded(state: BrowserState) {
+    private fun updatePriorityIfNeeded(state: BrowserState) = mainScope.launch {
         val currentSelectedTab = state.selectedTabId?.let { state.findTab(it) }
         val previousSelectedTab = state.findTab(previousHighestPriorityTabId)
         val currentEngineSession: EngineSession? = currentSelectedTab?.engineState?.engineSession
 
-        // We need to make sure, we alter the previousHighestPriorityTabId, after the session is linked.
+        // We need to make sure we alter the previousHighestPriorityTabId, after the session is linked.
         // So we update the priority on the engine session, as we could get actions where the tab
         // is selected but not linked yet, causing out sync issues,
         // when previousHighestPriorityTabId didn't call updateSessionPriority()
         if (currentEngineSession != null) {
-            previousSelectedTab?.engineState?.engineSession?.updateSessionPriority(DEFAULT)
+            mainScope.launch {
+                // check for existing form data here and if there is, set tab to DEFAULT
+                previousSelectedTab?.engineState?.engineSession?.checkForFormData()
+            }
+
             currentEngineSession.updateSessionPriority(HIGH)
             logger.info("Update the currentSelectedTab ${currentSelectedTab.id} priority to ${HIGH.name}")
-            logger.info("Update the previousSelectedTab ${previousSelectedTab?.id} priority to ${DEFAULT.name}")
             previousHighestPriorityTabId = currentSelectedTab.id
         }
     }
+
+    private fun updatePriorityToDefault(
+        context: MiddlewareContext<BrowserState, BrowserAction>,
+        tabId: String,
+        updatePriorityAfterMillis: Long,
+    ) {
+        // store and launch the new job related to the tabId
+        var updateJob: Job = waitScope.launch {
+            delay(updatePriorityAfterMillis)
+            context.store.dispatch(ContentAction.UpdatePriorityToDefaultAfterTimeoutAction(tabId))
```

Comment:
nit/out of curiosity: is this a common approach to dispatch an action inside an action that are privately interrelated to each other, i.e. `UpdatePriorityToDefaultAfterTimeoutAction` is (as far as i can imagine) internal and only dispatched whenever `CheckForFormDataAction` is triggered. As there is no external dispatching logic to this action, could we have just include the `UpdatePriorityToDefaultAfterTimeoutAction`'s lines inside middleware in  the `CheckForFormDataAction` case? No strong opinion here, and I may be wrong with my assumptions (i.e. the `UpdatePriorityToDefaultAfterTimeoutAction` is only internally handled by `CheckForFormDataAction`).

Reply:
I think it's an okay approach, it also helps with unit tests to be able to verify any action occurred. Since in either case, eventually the priority will result in `DEFAULT`.

Reply:
yes; makes sense. Just wondered whether it was smth we were avoiding or not :)